### ISSUE TITLE
consult-lsp-file-symbols not grouping by symbol kind

### ISSUE THREAD

**Issue by ashlineldridge** [gagbo/consult-lsp#23]
I'm having trouble getting `consult-lsp-file-symbols` to group symbol results by their kind. I have a fairly standard installation of `consult-lsp`:

```elisp
(use-package consult-lsp
  :custom
  (consult-lsp-marginalia-mode t))
```

By default, `consult-lsp-symbols` groups symbol results by their kind but `consult-lsp-file-symbols` displays its results ungrouped as shown in the screenshot below. Is this expected behavior or am I missing something? Ideally, I'd like to achieve a similar result as is shown in the screenshots in the project's README. This ungrouped behavior happens whether or not I enable Marginalia integration. I have not customized `consult-lsp-symbols-narrow` in any way.

Version information:
Emacs: 28.0.60
rust-analyzer: [2022-01-03](https://github.com/rust-analyzer/rust-analyzer/releases/tag/2022-01-03)
consult-lsp: [latest](https://github.com/gagbo/consult-lsp/commit/aaa9a31bc82259d743186c53d8b01f043c6fec13)

<img width="898" alt="Screen Shot 2022-01-04 at 10 48 14 am" src="https://user-images.githubusercontent.com/363071/147995952-3343ece1-ec24-45d8-85dd-d9dc02597d9d.png">

**Comment by gagbo** [comment]
Hello,

It is actually the intended behaviour for the `file-symbols` part. It's been discussed [in the PR](https://github.com/gagbo/consult-lsp/pull/13) that brought the support (although it's arguably hard to follow because there's also the marginalia discussion in the PR).

So the tl;dr for the rationale is : when you request `file-symbols`, there's value in letting all the symbols in the order they're defined in the file, so they are sorted by the line number where they appear (you can see that in the line column in the middle of the screenshots). You can still filter the candidates by typing into consult, and we've intentionally kept the fontification of the symbols, so that you can use it as a hint to guess the type.

If you want to narrow the list with categories, you should use the "global" `consult-lsp-symbols`. Arguably the usecase of single big main files like AoC code isn't really well covered by `file-symbols`, it's more meant to help in module files

**Comment by ashlineldridge** [comment]
@gagbo, thanks for the explanation.

That's good to know. I thought that I'd either misconfigured something or was seeing an odd interaction with another package.

Re: `consult-lsp-symbols` vs `consult-lsp-file-symbols`, `consult-lsp-symbols` only returns a very small number of symbols. For the example in the screenshot above, `consult-lsp-symbols` only returns 3 symbols (`impl` blocks and functions are ignored, for example) whereas `consult-lsp-file-symbols` returns 21 symbols. I'm guessing this is due to the different requests being made to the language server - i.e., `workspace/symbol` vs `textDocument/documentSymbol`?

Since `consult-lsp-file-symbols` returns the "complete" list of symbols, is there a recommended way of configuring consult-lsp so that these symbols are categorized?

**Comment by gagbo** [comment]
`textDocument/documentSymbol` requires in the request to be [scoped to a single document URI](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#textDocument_documentSymbol), that's why the server (rust-analyzer I'd guess here) allows itself to send more detailed answer to this request. So `consult-lsp-file-symbols` returns a "more complete" list of symbols only if the scope you're interested in is exclusively one file.

With that being said, to categorize the symbols in the request, the easiest way to get the grouping back would be to add another function with the necessary extra argument :
```lisp
(defun consult-lsp-file-symbols-grouped ()
  "Search symbols defined in current file in a manner similar to `consult-line'."
  (interactive)
  ;; Extra (require) to ensure all the utility functions are accessible
  (require 'consult-lsp)
  (consult--read
   (consult--with-increased-gc (consult-lsp--file-symbols-candidates))
   :prompt "Go to symbol: "
   :annotate (funcall consult-lsp-file-symbols-annotate-builder-function)
   :require-match t
   :sort nil
   :history '(:input consult--line-history)
   :category 'consult-lsp-file-symbols
   :lookup #'consult--line-match
   ;; Extra `:group` keyword argument to the file-symbols call
   :group (consult--type-group consult-lsp-symbols-narrow)
   :narrow (consult--type-narrow consult-lsp-symbols-narrow)
   :state (consult--jump-state)))
```

Now that I think about it, it might be a good candidate for a prefix argument for `consult-lsp-file-symbols`, when I get some time to tweak this I'll do it (feel free to make a PR if you want to). In either case, adding this extra definition in your `init.el` should work.

**Comment by ashlineldridge** [comment]
Thanks, @gagbo.

I'm happy to raise a PR. Is something like the following what you were thinking? Are you concerned with `consult-lsp-symbols` and `consult-lsp-file-symbols` having a different API?

```elisp
(defun consult-lsp-file-symbols ()
  "Search symbols defined in current file in a manner similar to `consult-line'. When prefix ARG is
   present symbols are grouped by kind otherwise they are returned in the order they appear in the file."
  (interactive "P")
  (consult--read
   (consult--with-increased-gc (consult-lsp--file-symbols-candidates))
   :prompt "Go to symbol: "
   :annotate (funcall consult-lsp-file-symbols-annotate-builder-function)
   :require-match t
   :sort nil
   :history '(:input consult--line-history)
   :category 'consult-lsp-file-symbols
   :lookup #'consult--line-match
   :narrow (consult--type-narrow consult-lsp-symbols-narrow)
   ;; Group results by symbol kind if prefix argument was specified.
   :group (when arg (consult--type-group consult-lsp-symbols-narrow))
   :state (consult--jump-state)))
```

**Comment by gagbo** [comment]
You're missing the `arg` in the argument list of `consult-lsp-file-symbols`, but otherwise that's what I was thinking of yeah. (And style wise, I like having a blank line between the description and the extra sentence about what arguments do; also renaming `arg` to `group-results` would be clearer probably).

I don't think it's a problem that the prefix arg does something different here and for `consult-lsp-{symbols,diagnostics}`; saying "prefix arg is all-workspaces" for a command that exclusively lists symbols in a file makes no sense, so the prefix must do something else for this command. And it looks like a good candidate :)

**Comment by ashlineldridge** [comment]
Thanks for the feedback. I'll raise a PR with the changes you mentioned and close this issue.
